api/layer/system_object.go

```diff
@@ -0,0 +1,171 @@
+	"github.com/nspcc-dev/neofs-s3-gw/api/cache"
+	"github.com/nspcc-dev/neofs-s3-gw/api/data"
+	"github.com/nspcc-dev/neofs-s3-gw/api/errors"
+	apiErrors "github.com/nspcc-dev/neofs-s3-gw/api/errors"
+	"go.uber.org/zap"
+)
+
+func (n *layer) putSystemObject(ctx context.Context, p *PutSystemObjectParams) (*object.Object, error) {
+	versions, _, err := n.headSystemVersions(ctx, p.bktInfo, p.objName)
+	if err != nil && !errors.IsS3Error(err, errors.ErrNoSuchKey) {
+		return nil, err
+	}
+	idsToDeleteArr := updateCRDT2PSetHeaders(p.metadata, versions, false) // false means "last write wins"
+
+	attributes := make([]*object.Attribute, 0, 3)
+
+	filename := object.NewAttribute()
+	filename.SetKey(objectSystemAttributeName)
+	filename.SetValue(p.objName)
+
+	createdAt := object.NewAttribute()
+	createdAt.SetKey(object.AttributeTimestamp)
+	createdAt.SetValue(strconv.FormatInt(time.Now().UTC().Unix(), 10))
+
+	versioningIgnore := object.NewAttribute()
+	versioningIgnore.SetKey(attrVersionsIgnore)
+	versioningIgnore.SetValue(strconv.FormatBool(true))
+
+	attributes = append(attributes, filename, createdAt, versioningIgnore)
+
+	for k, v := range p.metadata {
+		attr := object.NewAttribute()
+		attr.SetKey(p.prefix + k)
+		if p.prefix == tagPrefix && v == "" {
+			v = tagEmptyMark
+		}
+		attr.SetValue(v)
+		attributes = append(attributes, attr)
+	}
+
+	raw := object.NewRaw()
+	raw.SetOwnerID(p.bktInfo.Owner)
+	raw.SetContainerID(p.bktInfo.CID)
+	raw.SetAttributes(attributes...)
+
+	ops := new(client.PutObjectParams).WithObject(raw.Object()).WithPayloadReader(bytes.NewReader(p.payload))
+	oid, err := n.pool.PutObject(ctx, ops, n.BearerOpt(ctx))
+	if err != nil {
+		return nil, err
+	}
+
+	meta, err := n.objectHead(ctx, p.bktInfo.CID, oid)
+	if err != nil {
+		return nil, err
+	}
+
+	if p.payload != nil {
+		meta.ToV2().SetPayload(p.payload)
+	}
+
+	if err = n.systemCache.Put(cache.SystemObjectKey(p.bktInfo.Name, p.objName), meta); err != nil {
+		n.log.Error("couldn't cache system object", zap.Error(err))
+	}
+
+	for _, id := range idsToDeleteArr {
+		if err = n.objectDelete(ctx, p.bktInfo.CID, id); err != nil {
+			n.log.Warn("couldn't delete system object",
+				zap.Stringer("version id", id),
+				zap.String("name", p.objName),
+				zap.Error(err))
+		}
+	}
+
+	return meta, nil
+}
+
+func (n *layer) getSystemObject(ctx context.Context, bktInfo *data.BucketInfo, objName string, withPayload bool) (*object.Object, error) {
+	if meta := n.systemCache.Get(cache.SystemObjectKey(bktInfo.Name, objName)); meta != nil {
+		return meta, nil
+	}
+
+	versions, lastMeta, err := n.headSystemVersions(ctx, bktInfo, objName)
+	if err != nil {
+		return nil, err
+	}
+
+	if withPayload {
+		buf := new(bytes.Buffer)
+		p := &getParams{
+			Writer: buf,
+			cid:    bktInfo.CID,
+			oid:    versions.getLast().ID,
+		}
+		lastMeta, err = n.objectGet(ctx, p)
+		if err != nil {
+			return nil, err
+		}
+		lastMeta.ToV2().SetPayload(buf.Bytes())
```

Comment:
It seems we can get object without `PayloadWriter` and payload will be filled in response. But we need define another `objectGet` method in `api/layer/object.go` to get rid of the `WithPayloadWriter(w)` param.